### DISCUSSION TITLE
What keeps us from basing the code on c++14 or c++17?

### DISCUSSION THREAD

**Discussion by Naktakala** [chi-tech/chi-tech#121] [Q&A] [answered]
C++20 is obviously very new and I doubt many know how to use the features. It is also risky to assume that HPCs will have the necessary frameworks in place for C++20 compilers. With this in mind I was wondering if someone can add some insights into what difficulties we face if we were to attempt c++14 or c++17 compliance?

**Comment by wdhawkins** [comment]
I would strongly suggest we at least move to C++14.  I also don't see any reason not to move to C++17.  I've been using it for various projects, and compiler support is fairly robust at this point.  I doubt we'd run into many compilers that didn't support the standard (nvcc and xlc are the two exceptions I've encountered so far, but we can work around those minor irritations if need be).  Besides, structured binding is a beautiful thing and makes 17 almost a must have for me nowadays.
 
I would hesitate on 20.  Not all compilers will have mature support for 20 yet.

**Comment by ragusa** [comment, marked as answer]
I agree with Daryl.

**Comment by dominiccaron** [comment]
I second the movement to at least C++14, ideally C++17, for the reasons mentioned above.

**Comment by Naktakala** [comment]
I realized that I've been loosely talking about "compliance" but there are actually two things: "compliance" and "target standard". I think we've got a long way to go to make everything truly c++17 compliant, let alone c++14 and c++11, but I agree with everyone that we can go to c++17 as a "target standard".

@jthano Eager to get your inputs.

**Comment by jthano** [comment]
Unfortunately, I have not had much motivation to move beyond c++11. The only feature I've commonly encountered from c++14 is std::make_unique

**Comment by Naktakala** [comment]
Ok then! Moving to c++17. Let us see where it goes.
